### PR TITLE
Cnos Restapi Module

### PR DESCRIPTION
**Pull request by arunktele**
##### SUMMARY
Creating the Cnos Restapi module

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/network/cnos/cnos_restapi.py

##### ANSIBLE VERSION
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/usr/share/my_modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/network/cnos/cnos_restapi.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+version_added: "2.5"
+
+description:
+  - "Provides access to the RESTAPI's on the Lenovo CNOS switches.
+     RESTAPI interface on CNOS fetches data and configures the
+     Lenovo CNOS switches. "
+
+options:
+    transport:
```
Comment (gundalow) [inline lib/ansible/modules/network/cnos/cnos_restapi.py]: Please change this to 
```
use_ssl:
  description:
  type: bool
  default: true
```

File: lib/ansible/modules/network/cnos/cnos_restapi.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  description: Success or failure message
+  returned: always
+  type: string
+  sample: "RESTAPI PUT /nos/api/cfg/telemetry/bst/feature is successful"
+'''
+
+import json
+import ast
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import open_url
```
Comment (gundalow) [inline lib/ansible/modules/network/cnos/cnos_restapi.py:124]: Please use `get_url` which will give you built in error-checking

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cnos/cnos_restapi.py (hunk 1 above)
Comment (arunktele) [reply lib/ansible/modules/network/cnos/cnos_restapi.py]: I have fixed this.

File: lib/ansible/modules/network/cnos/cnos_restapi.py (hunk 2 above)
Comment (arunktele) [reply lib/ansible/modules/network/cnos/cnos_restapi.py:124]: The cnos restapi requires handling 401 error. Unfortunately the url library supresses this type of error. I had experimented with fetch_url and figured that it cannot be used, get_url cannot be used for this purpose to my knowledge. So I cannot fix this,

**Comment by sivel** [comment]
I've requested a review from myself to do some additional research and discussion here. We have recently discussed that we won't accept generic API modules. This module is a generic API module.  I'll need to clarify this stance before proceeding.

**Comment by sivel** [comment]
Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we have recently had a discussion on acceptable module scope and are not sure if we want this feature in the program, and I don't want this to seem confrontational.  Our reasons for this are:

   * Modules should have a concise and well defined functionality.  Basically, follow the UNIX philosophy of doing one thing well.
   * Modules should not require that a user know all the underlying options of an api/tool to be used.  For instance, if what the legal values for a required module parameter cannot be documented that's a sign the module would be rejected

Due to this, we believe this module does not meet the guidelines, as it requires the user to have full knowledge of the API for this to be used.

If you wish to submit multiple modules that deal explicitly in individual facets of the API, such as something like `cnos_interface` or `cnos_vlan` that explicitly list all configuration options as top level arguments to the module, we will be happy to review them for inclusion.

Thank you once again for this and your interest in Ansible!

**Pull request closed without merging**